### PR TITLE
Close on graceful #2749

### PR DESCRIPTION
**Pull request by gregw**
#2749
Add Connection:close on graceful shutdown responses

### CODE REVIEW

**Comment by gregw** [comment]
This PR has two implementations - in 01dbb44 the implementation uses a new field in `Response`, which makes a tiny cost for all requests, even if graceful shutdown is never used.   159f35d modifies that to use the `HttpChannel.Listener` mechanism, which results in a little higher cost, but only if the StatisticsHandler is used.

However, there is a problem with this later approach as the `onResponseBegin` method modifies the meta data from `request.getHttpChannel().getCommittedMetaData()`, which might actually be a static meta data supplied for a 4xx response!  Not sure how to work around that?

**Comment by gregw** [comment]
@sbordet also this approach only works for Http1.   Can we think of a mechanism that will also make http2 connections go away once the last response is sent?

**Comment by gregw** [comment]
@sbordet for http2, we filter out `Connection:close` headers because they are not allowed.  Perhaps we should note them and if the number of active streams is 0, send a go-away?

**Comment by gregw** [comment]
@sbordet can you look at this one

**Comment by gregw** [comment]
Wow, this is  getting bigger than Ben Hur! 
I've not added an `isShutdown` method and abstracted that into a `Graceful.Shutdown` utility class, so the `connection:close` will only be added if the connector is shutdown!

@janbartel can you also check the fix I made to the maven connector delegate to fix a very unlikely race condition...

**Comment by gregw** [comment]
I've removed the WIP label, as I think any closing of HTTP/2 connections after `Connection:close` should be handled in another issue.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/AbstractHandlerContainer.java [hunk 1]
```diff
@@ -135,4 +144,59 @@ public void setServer(Server server)
...
+            for (Future<Void> future: futures)
+            {
+                try
+                {
+                    if (!future.isDone())
+                        future.get(Math.max(1L,stop_by-System.currentTimeMillis()),TimeUnit.MILLISECONDS);
+                }
+                catch (Exception e)
+                {
+                    if (!future.isDone() && future instanceof Callback)
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/AbstractHandlerContainer.java]: Calling `shutdown()` may return a `Callback`? If so, where is the callback succeeded?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/AbstractHandlerContainer.java [hunk 2]
```diff
@@ -135,4 +144,59 @@ public void setServer(Server server)
...
+                        mex = new MultiException();
+                    mex.add(e);
+                }
+            }
+        }
+
+        // Cancel any shutdowns not done
+        for (Future<Void> future: futures)
+            if (!future.isDone())
+                future.cancel(true);
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/AbstractHandlerContainer.java]: If the future is a `Callback`, calling `cancel()` also fails the callback?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java [hunk 3]
```diff
@@ -117,6 +128,25 @@ public void onComplete(AsyncEvent event) throws IOException
             }   
         }
     };
+    
+    private final HttpChannel.Listener _shutdownListener = new HttpChannel.Listener()
+    {
+        @Override 
+        public void onResponseBegin(Request request)
+        {
+            if (_shutdown.isShutdown())
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java]: Do we need this `if` statement if we only add `Connection: close` when the connector is shutting down?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java [hunk 4]
```diff
@@ -117,6 +128,25 @@ public void onComplete(AsyncEvent event) throws IOException
...
+            {
+                Connector connector = request.getHttpChannel().getConnector();
+                if (connector instanceof Graceful && ((Graceful)connector).isShutdown())
+                {
+                    MetaData.Response response = request.getHttpChannel().getCommittedMetaData();
+
+                    if (!response.getFields().contains(HttpHeader.CONNECTION,"close"))
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java]: Use `HttpHeaderValue.CLOSE`?

Comment (sbordet) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java]: Can't we add this logic to the `HttpGenerator` instead here?
The `HttpGenerator` will ask whether the channel is shutting down, and will generate a `Connection: close`.

Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/StatisticsHandler.java]: Please add testcase where a servlet does `resp.setHeader("Connection", "CLOSE")`

**Comment by sbordet** [comment]
@gregw I would rethink this PR without having to use the `StatisticsHandler`.

The `HttpGenerator` is where we decide what headers we need to add depending on various states. The fact that the connector is shutting down is just another of the states that the generator should check to add `Connection: close`.

For HTTP/2 we need to stop accepting new requests, so we want to reset them immediately as soon as we detect that the connector is shutting down. Then we want to finish the outstanding requests (waiting at most some stop timeout), and then we want to send the `GOAWAY`.
I'll have a go at implementing this.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/AbstractHandlerContainer.java (hunk 2 above)
Comment (gregw) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/AbstractHandlerContainer.java]: It calls fail(throwable) above just to record what the exception is.

**Comment by gregw** [comment]
@joakime @sbordet Updated to do the shutdown check in HttpConnection as it calls the HttpGenerator.  If the connector is shutdown, then the generator is told that to not be persistent.  Also added a check for shutdown after commit so we can close anyway.

**Comment by gregw** [comment]
@joakime @sbordet re-review please!

**Comment by sbordet** [comment]
@gregw I like the new solution better, but please check that we don't overwrite `_persistent` during generation, in particular in `HttpGenerator.generateHeaders()` if we have a `Connection: Keep-Alive` header.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java [hunk 5]
```diff
@@ -644,7 +626,7 @@ private void generateHeaders(MetaData info,ByteBuffer header,ByteBuffer content,
         HttpField transfer_encoding=null;
         boolean http11 = info.getHttpVersion() == HttpVersion.HTTP_1_1;
         boolean close = false;
-        _trailers = http11?info.getTrailerSupplier():null;
+        _trailers = info.getHttpVersion() == HttpVersion.HTTP_1_1?info.getTrailerSupplier():null;
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpGenerator.java]: Revert this change, `http11` has been computed 2 lines above.
